### ISSUE TITLE
Mask-RCNN outputs

### ISSUE THREAD

**Issue by tersekmatija** [openvinotoolkit/openvino_contrib#253]
Hey,

I've managed to convert the Mask-RCNN, but I am having some issues understanding the outputs as they differ from the torchvision model.

Current outputs are: [DetectionOutput_647', 'Sigmoid_733', 'TopK_.0']. With `box_detections_per_img=25` and `rpn_post_nms_top_n_test=100`, their shapes are `(1, 1, 25, 7)`, `(25, 91, 28, 28)`, and `(100,)`, respectively. 

As far as I could understand, the DetectionOutput_647 contains `box_detections_per_img` boxes, where first element is either 0 or -1 (indicating last detected box which can appear sooner, depending on the NMS thresholds I assume). So, one row is:
```
[_, class_number, probability, x1, y1, x2, y2]
```

The Sigmoid_733 then contains 28x28 masks for ROI of each of the 25 boxes? And then 91 corresponds to the class label?

And what the TopK_.0 tells us?

Thanks!

**Comment by dkurt** [comment]
Hi! Sorry for confusion. Will add a code example with output parsing.

In short words, boxes output is `[batch_id, class_number, probability, left, top, right, bottom]` where coordinates are relative and should be multiplied to image shape. Masks output is `[box, class_id, height, width]` and contains masks probabilities for every class. So when you parse detections, get a corresponding class mask and resize from 28x28 to bounding box.

**Comment by tersekmatija** [comment]
Okay, thanks @dkurt ! I assumed so. One more question. I've noticed that when obtaining the segmentation mask, I have to use the index that is equal to the class label. So, Mask R-CNN pretrained on COCO has 91 classes. And the mask outputs second axis has length 91. I'd assume that for person (class 1) I'd have to access [box, **0**, :, :], so the index of the mask would be `class - 1`. That way class 1 -> 0, class 91 -> 90, and indices are OK and no Out of range error for the last class occurs. 

By testing, however, I've noticed that the correct mask is at the index `class` (at least I think so). I've tested for person (1) and mask at `class - 1` (0) seems to be completely wrong. For other classes masks at `class` and `class-1` are pretty similar, but the one at `class` has higher probabilities and is more certain. Is that the expected behavior or am I missing something? In case this is expected, how to handle the last class?

**Comment by tersekmatija** [comment]
Have you managed to check this @dkurt ?

**Comment by dkurt** [comment]
@tersekmatija, sorry for delayed response. Will take a look on this today.

**Comment by dkurt** [comment]
@tersekmatija, Please check a sample from https://github.com/openvinotoolkit/openvino_contrib/pull/266

The classes start from `__background__` so for public model `person` has class id 1. No need for subtraction. Maybe your private model has different classes annotation? If you don't have a background class at 0th position, can you manually modify `.xml` file and replace `background_label_id="0"` to `background_label_id="999"`. If that's the case - I will try to manage automatic `background_label_id` extraction from the model.

**Comment by tersekmatija** [comment]
Hey @dkurt ,

thanks for the example! I was parsing the outputs the same way (so person on index `1`, and I ignore index `0`). And I used the default model from torchvision, so my demo was correct. 😄 

What I was confused with is that the torchvision model is pre-trained on 91 classes (person - hair brush), where as the outputs of the OpenVINO model have only 90 classes by that parsing logic (`background + 90 = 91` instead of `background + 91 = 92`). I was wondering what you do with the last class (hair brush), that would correspond to the 92th element in the OpenVINO outputs. I noticed that in your example you just drop it and not use it at all?

**Comment by dkurt** [comment]
@tersekmatija, Do you mean `toothbrush`? Now OpenVINO predicts all 91 classes (background + 90). There are 90 real classes, not 91. Masks tensor shape is `100x91x298x28`. Maybe I misunderstood your question?

<img src="https://user-images.githubusercontent.com/25801568/146183585-066b97a9-cc23-4885-975b-60d6dad53318.png" width="256">

**Comment by tersekmatija** [comment]
You understood correctly @dkurt , but we seem to be confused on the number of total classes. I thought that 91 classes means 91 actual classes (like here: https://github.com/amikelive/coco-labels/blob/master/coco-labels-paper.txt) + 1 class background. Seems that the toothbrush is actually the last class then, and that hair brush was removed from the COCO.
